tracing-appender/src/compression.rs

```diff
@@ -0,0 +1,154 @@
+use flate2::Compression;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) enum GzipCompressionLevelLiteral {
+    None,
+    Fast,
+    Best,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) enum GzipCompressionLevelNumerical {
+    Level0,
+    Level1,
+    Level2,
+    Level3,
+    Level4,
+    Level5,
+    Level6,
+    Level7,
+    Level8,
+    Level9,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) enum GzipCompressionLevel {
+    Literal(GzipCompressionLevelLiteral),
+    Numerical(GzipCompressionLevelNumerical),
+}
+
+/// Defines a conversion between `CompressionOption` and `flate2::Compression`
+impl Into<Compression> for GzipCompressionLevel {
+    fn into(self) -> Compression {
+        match self {
+            GzipCompressionLevel::Literal(lit) => match lit {
+                GzipCompressionLevelLiteral::None => Compression::none(),
+                GzipCompressionLevelLiteral::Fast => Compression::fast(),
+                GzipCompressionLevelLiteral::Best => Compression::best(),
+            },
+            GzipCompressionLevel::Numerical(num) => match num {
+                GzipCompressionLevelNumerical::Level0 => Compression::new(0),
+                GzipCompressionLevelNumerical::Level1 => Compression::new(1),
+                GzipCompressionLevelNumerical::Level2 => Compression::new(2),
+                GzipCompressionLevelNumerical::Level3 => Compression::new(3),
+                GzipCompressionLevelNumerical::Level4 => Compression::new(4),
+                GzipCompressionLevelNumerical::Level5 => Compression::new(5),
+                GzipCompressionLevelNumerical::Level6 => Compression::new(6),
+                GzipCompressionLevelNumerical::Level7 => Compression::new(7),
+                GzipCompressionLevelNumerical::Level8 => Compression::new(8),
+                GzipCompressionLevelNumerical::Level9 => Compression::new(9),
+            },
+        }
+    }
+}
```

Comment:
nit, take it or leave it: i'm not totally sure if the `GzipCompressionLevelNumerical` enum is really necessary; it feels like it makes this code much more complex but isn't really necessary. what if we just defined this as
```suggestion
#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) enum GzipCompressionLevel {
    Literal(GzipCompressionLevelLiteral),
    Numerical(u32),
}

/// Defines a conversion between `CompressionOption` and `flate2::Compression`
impl Into<Compression> for GzipCompressionLevel {
    fn into(self) -> Compression {
        match self {
            GzipCompressionLevel::Literal(lit) => match lit {
                GzipCompressionLevelLiteral::None => Compression::none(),
                GzipCompressionLevelLiteral::Fast => Compression::fast(),
                GzipCompressionLevelLiteral::Best => Compression::best(),
            },
            GzipCompressionLevel::Numerical(num) => {
                assert!(num <= 9, "GZIP compression levels must be in the range 0-9");
                Compression::new(num)
            }
        }
    }
}
```

instead?

Since the only way for user code to construct a `GzipCompressionLevel` is via the `CompressionOption` type, I don't think we really have to worry about out of range compression levels ever being constructed, so I'm not sure if the enum is really necessary here.

Alternatively, if you think it's really important to make invalid numbers unrepresentable, we could do something like this to at least simplify the `Into<Compression>` impl a bit:
```suggestion
#[derive(Debug, Clone, Eq, PartialEq)]
#[repr(u32)]
pub(crate) enum GzipCompressionLevelNumerical {
    Level0 = 0,
    Level1 = 1,
    Level2 = 2,
    Level3 = 3,
    Level4 = 4,
    Level5 = 5,
    Level6 = 6,
    Level7 = 7,
    Level8 = 8,
    Level9 = 9,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) enum GzipCompressionLevel {
    Literal(GzipCompressionLevelLiteral),
    Numerical(GzipCompressionLevelNumerical),
}

/// Defines a conversion between `CompressionOption` and `flate2::Compression`
impl Into<Compression> for GzipCompressionLevel {
    fn into(self) -> Compression {
        match self {
            GzipCompressionLevel::Literal(lit) => match lit {
                GzipCompressionLevelLiteral::None => Compression::none(),
                GzipCompressionLevelLiteral::Fast => Compression::fast(),
                GzipCompressionLevelLiteral::Best => Compression::best(),
            },
            GzipCompressionLevel::Numerical(num) => Compression::new(num as u32)
        }
    }
}
```